src/tls/mod.rs

```diff
@@ -0,0 +1,575 @@
+    ///     ))
+    ///     .build();
+    /// # Ok::<(), isahc::Error>(())
+    /// ```
+    pub fn identity(mut self, identity: Identity) -> Self {
+        self.identity = Some(identity);
+        self
+    }
+
+    /// Set the minimum allowed protocol version for secure connections.
+    ///
+    /// If specified, the client will attempt to negotiate secure connections
+    /// using the given protocol version or newer if possible. If a server only
+    /// supports protocols older than the set minimum then the connection will
+    /// be terminated with a
+    /// [`ConnectionFailed`][crate::error::ErrorKind::ConnectionFailed] error.
+    ///
+    /// The default value if not set depends on the TLS backend Isahc is
+    /// configured to use. With most backends on most modern systems the default
+    /// will be TLS v1.2, but if you want to be certain then you may want to set
+    /// a minimum version explicitly.
+    pub fn min_version(mut self, version: ProtocolVersion) -> Self {
+        self.min_version = Some(version);
+        self
+    }
+
+    /// Set a maximum allowed protocol version for secure connections.
+    ///
+    /// If specified, the client will attempt to negotiate secure connections
+    /// using the given protocol version or older if possible. If a server only
+    /// supports protocols newer than the set maximum then the connection will
+    /// be terminated with a
+    /// [`ConnectionFailed`][crate::error::ErrorKind::ConnectionFailed] error.
+    ///
+    /// By default no maximum version is enforced, however the newest version
+    /// that can be used will be dependent on what newest protocols are
+    /// supported by the underlying TLS engine.
+    pub fn max_version(mut self, version: ProtocolVersion) -> Self {
+        self.max_version = Some(version);
+        self
+    }
+
+    /// Set a list of ciphers or cipher suites to allow.
+    ///
+    /// The list of valid cipher names is dependent on the underlying SSL/TLS
+    /// backend in use. You can find an up-to-date list of potential cipher names
+    /// at <https://curl.haxx.se/docs/ssl-ciphers.html>.
+    ///
+    /// The default is unset and will result in using whatever ciphers the
+    /// configured TLS backend chooses to enable by default.
+    ///
+    /// # Warning
+    ///
+    /// Not all cipher suites are created equal, as many have been marked as
+    /// being insecure over time. This method allows you to enable potentially
+    /// insecure cipher suites that are not enabled by default and may introduce
+    /// security vulnerabilities into your application.
+    pub fn ciphers<I, T>(mut self, ciphers: I) -> Self
+    where
+        I: IntoIterator<Item = T>,
+        T: AsRef<str>,
+    {
+        let mut iter = ciphers.into_iter();
+
+        // If an empty list is provided, reset to default. Otherwise build up a
+        // string in curl format containing the cipher names.
+        if let Some(first) = iter.next() {
```

Comment:
What's the different with `ciphers.into_iter().join(":")`?

Reply:
When an empty list is provided, we're simply not setting any ciphers, and allowing the TLS engine to choose whichever ciphers they think is best. If we _actually_ set it to an empty list though, some TLS engines will interpret this as, "No ciphers are allowed" and basically reject all connections.

Reply:
How about:

```rust
let ciphers = ciphers.into_iter().join(":");
self.ciphers = if ciphers.is_empty() {None} else {Some(ciphers)}
```

Looks nicer to me.